packages/frames.js/src/core/formatUrl.ts

```diff
@@ -0,0 +1,121 @@
+// Format function modified from nodejs
+// Copyright Joyent, Inc. and other Node contributors.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a
+// copy of this software and associated documentation files (the
+// "Software"), to deal in the Software without restriction, including
+// without limitation the rights to use, copy, modify, merge, publish,
+// distribute, sublicense, and/or sell copies of the Software, and to permit
+// persons to whom the Software is furnished to do so, subject to the
+// following conditions:
+//
+// The above copyright notice and this permission notice shall be included
+// in all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS
+// OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+// MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN
+// NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM,
+// DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR
+// OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE
+// USE OR OTHER DEALINGS IN THE SOFTWARE.
+
+import type { UrlObject } from "url";
+import type { ParsedUrlQuery } from "querystring";
+
+// function searchParamsToUrlQuery(searchParams: URLSearchParams): ParsedUrlQuery {
+//   const query: ParsedUrlQuery = {};
+//   searchParams.forEach((value, key) => {
+//     if (typeof query[key] === "undefined") {
+//       query[key] = value;
+//     } else if (Array.isArray(query[key])) {
+//       (query[key] as string[]).push(value);
+//     } else {
+//       query[key] = [query[key] as string, value];
+//     }
+//   });
+//   return query;
+// }
+
+function stringifyUrlQueryParam(param: unknown): string {
+  if (
+    typeof param === "string" ||
+    (typeof param === "number" && !isNaN(param)) ||
+    typeof param === "boolean"
+  ) {
+    return String(param);
+  } else {
+    return "";
+  }
+}
+
+function urlQueryToSearchParams(urlQuery: ParsedUrlQuery): URLSearchParams {
+  const result = new URLSearchParams();
+  Object.entries(urlQuery).forEach(([key, value]) => {
+    if (Array.isArray(value)) {
+      value.forEach((item) => result.append(key, stringifyUrlQueryParam(item)));
+    } else {
+      result.set(key, stringifyUrlQueryParam(value));
+    }
+  });
+  return result;
+}
+
+// function assign(
+//   target: URLSearchParams,
+//   ...searchParamsList: URLSearchParams[]
+// ): URLSearchParams {
+//   searchParamsList.forEach((searchParams) => {
+//     Array.from(searchParams.keys()).forEach((key) => target.delete(key));
+//     searchParams.forEach((value, key) => target.append(key, value));
+//   });
+//   return target;
+// }
+
+const slashedProtocols = /https?|ftp|gopher|file/;
+
+export function formatUrl(urlObj: UrlObject) {
```

Comment:
why is this function necessary? isn't `new URL({})` already handling formatting?

Reply:
types don't line up

Reply:
new URL doesn't allow passing in a plain object for query params for example, amongst other things